spec/core/ics-004-channel-and-packet-semantics/UPGRADES.md

```diff
@@ -285,17 +303,33 @@ function chanUpgradeTry(
     portIdentifier: Identifier,
     channelIdentifier: Identifier,
     counterpartyChannel: ChannelEnd,
+    counterpartySequence: uint64,
     proposedUpgradeChannel: ChannelEnd,
     timeoutHeight: Height,
     timeoutTimestamp: uint64,
     proofChannel: CommitmentProof,
     proofUpgradeTimeout: CommitmentProof,
+    proofSequence: CommitmentProof,
     proofHeight: Height
 ) {
     // current channel must be OPEN or UPGRADE_INIT (crossing hellos)
     currentChannel = provableStore.get(channelPath(portIdentifier, channelIdentifier))
     abortTransactionUnless(currentChannel.state == OPEN || currentChannel.state == UPGRADE_INIT)
 
+    // get current sequence on this channel
+    // if the counterparty sequence is greater than the current sequence, we fast forward to the counterparty sequence
+    // so that both channel ends are using the same sequence for the current upgrade
+    // if the counterparty sequence is less than the current sequence, then either the counterparty chain is out-of-sync or
+    // the message is out-of-sync and we write an error receipt for that sequence to avoid a possible collision with previous upgrade attempt
+    currentSequence = provableStore.get(currentSequencePath(portIdentifier, channelIdentifier))
+    abortTransactionUnless()
+    if counterpartySequence > currentSequence {
```

Comment:
I think there is an issue here. If we keep it `>`, then the following scenario is possible. Chain `A` starts an upgrade with `seq=10` and moves to `UPGRADE_INIT`. A malicious relayer sends to `A` a `ChanUpgradeTry` with the `counterpartySequence` set to `10`, which will cancel the upgrade. In general, the `counterpartySequence` should be verified before updating the state (i.e., `errorPath`). However, this scenario could happen even with correct relayers, if both chain `A` and the counterparty (chain `B`) start an upgrade with the same sequence `10`. 

Alternatively, if we change `>` to `>=`, then the following scenario is possible. Chain `A` just finished executing `chanUpgradeAck` for sequence `10`, i.e., the upgrade is successful and `A` moves to `OPEN`. A relayer sends to `A`  a `ChanUpgradeTry` with the `counterpartySequence` set to `10`, which will trigger a new update for the same sequence.

Reply:
Would it be safer to bump the sequence on ACK and CONFIRM and then switch to `>=`?

I think it would also be nice to track the sequences in state so we don't allow relayers to pass in a sequence at all. But that may end up being too complicated if the individual sequences on each chain go out of sync. I tried to think a few times of how to do it, but there were scenarios in which one chain gets its sequence incremented a few times then initiating from the other chain would fail until the sequences were back in sync

Reply:
> Would it be safer to bump the sequence on ACK and CONFIRM and then switch to `>=`?

It will solve the scenario described above. I'm not sure though if it will not introduce other issues. For example, in CONFIRM the sequence needs to be bumped  before the sequenceProof is verified.

Reply:
What about the following approach that uses heights as sequence?

- Initially, each chain has its own `errorPath` set to `""`. 
- Chain `A` recv INIT -> `seq = currentHeightA`.
- Chain `B` recv TRY(sequenceProof) -> `seq = currentHeightB`, `counterpartySeq = currentHeightA`
- Chain `A` recv ACK(sequenceProof) -> `counterpartySeq = currentHeightB`

Note that 
- INIT does not require a `sequence`.
- CONFIRM does not require `sequenceProof` since the `proofHeight` can be compared to the `counterpartySeq` (i.e., must be `>=`).

`restoreChannel()` is called either on TRY or on ACK. When called, the  `errorPath` is set to the `counterpartySeq`.

Reply:
I believe I've solved this issue in the following changes:

1. Allow upgrade on TRY if counterparty sequence is `>=` to our sequence. We then will set the counterparty sequence as our sequence on TRY
2. If counterparty sequence is `<` our sequence, write our sequence into the error receipt so that counterparty can be "caught up" efficiently and without relayer input
3. Increment sequences on ACK, CONFIRM, and error cases
4. Do not increment sequence on timeout

With these changes, I don't even think it is possible for sequences to go out of sync. But i have the logic to handle it in case it is.